### PR TITLE
chore(asm): enable iast for threat tests

### PR DESCRIPTION
**Pull request by christophe-papazian**
## Motivation
To increase test coverage and problem detection, we need to be able to test several features at once.

## This PR
- add, for each threat test suite, a second run with IAST enabled
- change the way the threat tests use the tracer instrumentation, by directly using `ddtrace.auto` instead of calling specific patching methods. This makes the whole test suites closer to a real application
- fix a small regression in response header report for fastapi that was detected in those new tests
- fix a small (unreleased) bug in Django 2.2 not working with IAST 
- improve `repr` for `interface` fixture to allow more user friendly reports in case of CI failure
- relax some header test for api security as the tracer may add additional headers in the response.
- fix an (unreleased) bug in exploit prevention blocking mechanism

## Revealed Issues

- Exploit Prevention and IAST are not working properly together for now.
- Root span is not reported in the span list by the tracer object with recent versions of fastAPI (probably a test only problem)

APPSEC-53115

## Checklist

- [x] Change(s) are motivated and described in the PR description
- [x] Testing strategy is described if automated tests are not included in the PR
- [x] Risks are described (performance impact, potential for breakage, maintainability)
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [x] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.

## Reviewer Checklist

- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Description motivates each change
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] Release note makes sense to a user of the library
- [x] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/appsec/_iast/taint_sinks/header_injection.py [hunk 1]
```diff
@@ -47,8 +47,12 @@ def _(m):
     @when_imported("django.http.response")
     def _(m):
         trace_utils.wrap(m, "HttpResponse.__setitem__", _iast_h)
-        trace_utils.wrap(m, "ResponseHeaders.__setitem__", _iast_h)
         trace_utils.wrap(m, "HttpResponseBase.__setitem__", _iast_h)
+        try:
+            trace_utils.wrap(m, "ResponseHeaders.__setitem__", _iast_h)
+        except AttributeError:
+            # no ResponseHeaders in django<3
+            pass
```
Comment (gnufede) [inline ddtrace/appsec/_iast/taint_sinks/header_injection.py:55]: I think we should do this block in a different PR and backport it to 2.7, 2.8, 2.9...

Comment (christophe-papazian) [reply ddtrace/appsec/_iast/taint_sinks/header_injection.py:55]: I can do that.

File: tests/appsec/contrib_appsec/utils.py [hunk 2]
```diff
@@ -1212,90 +1222,75 @@ def test_exploit_prevention(
...
+            else:
+                assert get_triggers(root_span()) is None
+                assert self.check_for_stack_trace(root_span) == []
+                assert get_tag("rasp.request.done") == endpoint
 
-    @pytest.mark.skip(reason="iast integration not working yet")
-    def test_iast(self, iast, interface, root_span, get_tag):
-        from ddtrace.appsec._iast.taint_sinks.command_injection import patch
-        from ddtrace.appsec._iast.taint_sinks.command_injection import unpatch
+    # @pytest.mark.skip(reason="iast integration not working yet")
```
[the remark this answers is not in the document]
Comment (christophe-papazian) [reply tests/appsec/contrib_appsec/utils.py:1278]: yes, that was commented, but I'll remove it

File: ddtrace/appsec/_iast/taint_sinks/header_injection.py (hunk 1 above)
Comment (avara1986) [reply ddtrace/appsec/_iast/taint_sinks/header_injection.py:55]: `trace_utils.wrap(m, "ResponseHeaders.__setitem__", _iast_h)` isn't in any release

Comment (christophe-papazian) [reply ddtrace/appsec/_iast/taint_sinks/header_injection.py:55]: Oh yes, it was merged only 3 weeks ago. OK, I let it in that PR then, thanks @avara1986

**Review by juanjux — APPROVED** [review]

File: tests/appsec/contrib_appsec/conftest.py [hunk 3]
```diff
@@ -1,30 +1,16 @@
-import unittest.mock
+import ddtrace.auto  # noqa: F401
 
-import pytest
 
-from ddtrace.settings.asm import config as asm_config
-from tests.utils import TracerSpanContainer
-from tests.utils import _build_tree
+# ensure the tracer is loaded and started first for possible iast patching
+print(f"ddtrace version {ddtrace.version.get_version()}")
```
Comment (romainkomorndatadog) [inline tests/appsec/contrib_appsec/conftest.py:5]: This is kind of weird?

File: tests/appsec/contrib_appsec/flask_app/app.py [hunk 4]
```diff
@@ -4,10 +4,15 @@
 from flask import request
 
 from ddtrace import tracer
+
+# from ddtrace.appsec.iast import ddtrace_iast_flask_patch
 import ddtrace.constants
 from tests.webclient import PingFilter
 
 
+# ddtrace_iast_flask_patch()
```
Comment (romainkomorndatadog) [inline tests/appsec/contrib_appsec/flask_app/app.py:13]: remove?
```suggestion
```

File: tests/appsec/contrib_appsec/conftest.py (hunk 3 above)
Comment (christophe-papazian) [reply tests/appsec/contrib_appsec/conftest.py:5]: But it works great.

Comment (christophe-papazian) [reply tests/appsec/contrib_appsec/conftest.py:5]: The print is not relevant, it's the import before of course that is.

Comment (romainkomorndatadog) [reply tests/appsec/contrib_appsec/conftest.py:5]: The print line is the line I commented on. Why are we printing something to stdout at module import time without even using a logger?

**Pull request merged**